README.md
# Christmas Spirit

For MC 1.16.4  

♫ Deck the mines with boughs of holly! ♫  
Wrap presents, eat holiday sweets, and decorate your Minecraft world with Christmas-themed features!

### This fork fixes an incompatibility bug with FakeName.
You may be impacted by this bug if your Minecraft Client crashes while using FakeName with the following error:

`io.netty.handler.codec.EncoderException: String too big (was 6 bytes encoded, max 5)`


changelog.txt
Build: 1.16.4-35.0.8 - Sun Oct 25 11:02:44 GMT 2021
	padlocks:
		Fixes wrapping presents from crashing the client when FakeName is present by removing formatting codes from player names internally.

=========
Build: 1.16.4-35.0.7 - Sun Nov 08 04:12:02 GMT 2020
	cpw:
		Some help in ExtensionPoint regarding DISPLAYTEST. Let's improve those tick rates!
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.6 - Sun Nov 08 03:30:37 GMT 2020
	cpw:
		Handle erroring during early mod construction phases, and actually report that into the error UI
		by doing armslength exception handling and propagation.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.5 - Sun Nov 08 01:48:22 GMT 2020
	cpw:
		Support backwards loading 1.16.3 mods in 1.16.4, because we are able to do that. Tweak loading a bit to be smarter about dependency errors of various kinds.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.4 - Sat Nov 07 20:10:44 GMT 2020
	David Quintana:
		Add FluidStack codec, and a test mod that verifies its behaviour matches the existing write/read logic.
		Add missing license headers.

=========
Build: 1.16.4-35.0.3 - Sat Nov 07 02:11:47 GMT 2020
	cpw:
		Update event bus version (should be 3.0.5)
		
		Tweak MDK to automatically populate recommended values for things like forge version, loader version and minecraft version.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.2 - Wed Nov 04 00:07:58 GMT 2020
	diesieben07:
		Workaround for modded attributes on vanilla entities logging warnings on vanilla clients
		
		Switch to a handler-based approach per packet class, simplify patch, add some docs
		
		Add license header and docs
		
		Switch to a network handler based approach
		
		Switch to a network handler based approach
		
		revert build gradle change
		
		Simplify Packet patch
		
		Add license header to IForgePacket
		
		Revert ForgeHooks changes
		
		Less hacky and more generic way to filter packets for vanilla connections
		
		_Actually_ Order SAS.
		It's early
		
		Order SAS
		
		simplify vanilla connection check
		
		Remove unneeded import change

=========
Build: 1.16.4-35.0.1 - Mon Nov 02 19:57:45 GMT 2020
	cpw:
		Fix early sorting bug that meant the file deduping was causing an error, to, well, not cause it, so we can get to an error screen.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.0 - Mon Nov 02 19:15:50 GMT 2020
	lex:
		1.16.4 Initial Update

=========
Build: 1.16.3-34.1.42 - Mon Nov 02 04:30:53 GMT 2020
	lex:
		Fix Biomes not properly copying over ids loaded from the save to the new Dynamic Registry.

=========
Build: 1.16.3-34.1.41 - Mon Nov 02 01:09:57 GMT 2020
	David Quintana:
		Introduce custom loader additions to the model data generators. (#7450)
		
		Currently implemented loaders:
		* OBJ
		* Composite
		* Multi-layer
		* Item layers (vanilla item/generated but with fullbright texture support)
		* Bucket
		* Separate Perspective

=========
Build: 1.16.3-34.1.40 - Mon Nov 02 00:52:12 GMT 2020
	cpw:
		actually rollback on error in registry events.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.3-34.1.39 - Sun Nov 01 22:09:32 GMT 2020
	cpw:
		Update modlauncher. fixes #7452
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

	cpw:
		Return client resources even if no data pack found on server. Closes #6225 (Pull request)
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

	tobias.hotz:
		Auto-detect ansi support for log4j2
		
		move to fml.loading subpackage
		
		Add license header

	cpw:
		Fix config parse failure causing crash in building exception message. fixes #7438
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.3-34.1.35 - Thu Oct 29 17:09:31 GMT 2020
	malte0811:
		Fix (NonNull)Lazy.Concurrent using a global lock (#7403)
		
		* Fix Lazy.Concurrent using a global lock, thus preventing multiple threads from resolving independent Lazy's at the same time
		* Use Lazy with an added null check to implement NonNullLazy

=========
Build: 1.16.3-34.1.34 - Thu Oct 29 00:12:09 GMT 2020
	David Quintana:
		Update forge mappings to 20201028-1.16.3 (#7435)

=========
Build: 1.16.3-34.1.33 - Tue Oct 27 16:01:02 GMT 2020
	35673674+alcatrazEscapee:
		Fix breaking changes from the Biome rename PR #7434 (#7439)

=========
Build: 1.16.3-34.1.32 - Mon Oct 26 23:52:47 GMT 2020
	sciwhiz12:
		Reimplement nametag distance attribute, fix attribute translation keys (#7387)

=========
Build: 1.16.3-34.1.31 - Mon Oct 26 22:07:02 GMT 2020
	35673674+alcatrazEscapee:
		Fix data pack biomes failing to load in SingleBiomeProvider (#7434)

=========
Build: 1.16.3-34.1.30 - Mon Oct 26 22:00:28 GMT 2020
	cpw:
		Add in proper handling of equals and hashcode for modjar urls. Fixes very slow loading on windows.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.3-34.1.29 - Mon Oct 26 12:51:24 GMT 2020
	cpw:
		Force 8.0.5 ML
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.3-34.1.28 - Mon Oct 26 11:32:40 GMT 2020
	cpw:
		Restore FingerprintViolationEvent with appropriate THIS IS F*CKING DEAD DELET WARNINGS.
		Also, notify in UI if using legacy JDK without capability to read out security data from mod jars.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.3-34.1.27 - Mon Oct 26 02:27:48 GMT 2020
	cpw:
		Rollback registries to vanilla state if something happens during loading, so subsequent activities can still run and not generate false reports..
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

	cpw:
		Add signature reporting for mods, using new signature capture from ModLauncher. Need to figure out
		how to reflect Minecraft's JAR signatures into here.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.3-34.1.25 - Tue Oct 20 19:32:40 GMT 2020
	lex:

src/main/java/com/tm/cspirit/block/BlockPresentWrapped.java
package com.tm.cspirit.block;

import com.tm.cspirit.init.InitSounds;
import com.tm.cspirit.init.InitTileEntityTypes;
import com.tm.cspirit.present.PresentConstructor;
import com.tm.cspirit.tileentity.TileEntityPresentWrapped;
import com.tm.cspirit.util.Location;
import com.tm.cspirit.util.helper.ItemHelper;
import com.tm.cspirit.util.helper.PresentHelper;
import com.tm.cspirit.util.helper.SoundHelper;
import com.tm.cspirit.util.helper.TimeHelper;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.ServerPlayerEntity;
import net.minecraft.fluid.FluidState;
import net.minecraft.inventory.ItemStackHelper;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.ActionResultType;
import net.minecraft.util.Hand;
import net.minecraft.util.NonNullList;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.BlockRayTraceResult;
import net.minecraft.util.math.shapes.ISelectionContext;
import net.minecraft.util.math.shapes.VoxelShape;
import net.minecraft.util.text.TextFormatting;
import net.minecraft.world.IBlockReader;
import net.minecraft.world.World;

import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class BlockPresentWrapped extends BlockPresentUnwrapped {

    public static final VoxelShape SHAPE = Optional.of(Block.makeCuboidShape(1, 0, 1, 15, 13, 15)).get();

    @Override
    public TileEntity createNewTileEntity(IBlockReader worldIn) {
        return InitTileEntityTypes.PRESENT_WRAPPED.get().create();
    }

    public static void spawnPresent(Location location, PresentConstructor constructor, ItemStack giftStack) {

        ItemStack stack = new ItemStack(constructor.getStyle().getBlock().asItem());

        constructor.toStack(stack);

        CompoundNBT nbt = ItemHelper.getNBT(stack);
        NonNullList<ItemStack> giftList = NonNullList.create();
        giftList.add(0, giftStack);
        ItemStackHelper.saveAllItems(nbt, giftList);

        ItemHelper.spawnStackAtLocation(location.world, location, stack);
    }

    @Override
    public boolean removedByPlayer(BlockState state, World world, BlockPos pos, PlayerEntity player, boolean willHarvest, FluidState fluid) {

        Location location = new Location(world, pos);
        TileEntity tileEntity = location.getTileEntity();

        if (!world.isRemote) {

            if (tileEntity instanceof TileEntityPresentWrapped) {

                TileEntityPresentWrapped present = (TileEntityPresentWrapped) tileEntity;

                boolean isAnybody = present.getConstructor().getToPlayerName().equalsIgnoreCase("anybody");
                boolean isToPlayer = player.getDisplayName().getString().equalsIgnoreCase(present.getConstructor().getToPlayerName());
                boolean isFromPlayer = player.getDisplayName().getString().equalsIgnoreCase(present.getConstructor().getFromPlayerName());

                if (isAnybody|| isToPlayer || isFromPlayer || player.isCreative()) {
                    spawnPresent(location, present.getConstructor(), present.getInventory().getStackInSlot(0));
                    location.setBlockToAir();
                }

                else present.getUnitName(player).printMessage(TextFormatting.RED, "This present belongs to " + present.getConstructor().getToPlayerName() + "! You can't pick it up!");
            }
        }

        return true;
    }
    @Override
    public ActionResultType onBlockActivated (BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockRayTraceResult result) {

        Location location = new Location(world, pos);
        TileEntity tileEntity = location.getTileEntity();

        if (!world.isRemote) {

            if (tileEntity instanceof TileEntityPresentWrapped) {

                TileEntityPresentWrapped present = (TileEntityPresentWrapped) tileEntity;
                Pattern pattern = Pattern.compile("[\u00A7].");
                Matcher matcher = pattern.matcher(player.getDisplayName().getString());
                String cleanedString = matcher.replaceAll("");

                if (player.isCrouching()) {
                    present.getUnitName(player).printMessage(TextFormatting.WHITE, "From: " + present.getConstructor().getFromPlayerName());
                    present.getUnitName(player).printMessage(TextFormatting.WHITE, "To: " + present.getConstructor().getToPlayerName());
                    present.getUnitName(player).printMessage(TextFormatting.WHITE, "Open on the " + TimeHelper.getFormattedDay(present.getConstructor().getActualDay()));
                }

                else if (cleanedString.equalsIgnoreCase(present.getConstructor().getToPlayerName()) || present.getConstructor().getToPlayerName().equalsIgnoreCase("anybody")) {

                    if (TimeHelper.getCurrentDay() >= present.getConstructor().getActualDay()) {

                        ItemStack stack = present.getInventory().getStackInSlot(0);

                        if (present.getConstructor().getFromPlayerName().equalsIgnoreCase("santa")) {
                            stack = PresentHelper.getSantaGiftedStack(player, present.getConstructor().getDay());
                        }

                        ItemHelper.spawnStack(world, pos.getX() + 0.5F, pos.getY() + 0.5F, pos.getZ() + 0.5F, stack);
                        location.setBlockToAir();

                        player.playSound(InitSounds.PRESENT_UNWRAP.get(), 1, 1);
                        SoundHelper.sendSoundToClient((ServerPlayerEntity) player, InitSounds.PRESENT_UNWRAP.get());
                    }

                    else {
                        present.getUnitName(player).printMessage(TextFormatting.RED, "You can't open this present yet!");
                        present.getUnitName(player).printMessage(TextFormatting.RED, "You must wait until the " + TimeHelper.getFormattedDay(present.getConstructor().getActualDay()) + "!");
                    }
                }

                else {
                    present.getUnitName(player).printMessage(TextFormatting.RED, "This present belongs to " + present.getConstructor().getToPlayerName() + "! You can't open it!");
                }
            }
        }

        return ActionResultType.SUCCESS;
    }

    @Override
    public VoxelShape getShape (BlockState state, IBlockReader worldIn, BlockPos pos, ISelectionContext context){
        return SHAPE;
    }

    @Override
    public VoxelShape getCollisionShape (BlockState state, IBlockReader world, BlockPos pos, ISelectionContext context) {
        return SHAPE;
    }
}


src/main/java/com/tm/cspirit/client/gui/ScreenPresentUnwrapped.java
package com.tm.cspirit.client.gui;

import com.mojang.blaze3d.matrix.MatrixStack;
import com.tm.cspirit.client.gui.base.ButtonRect;
import com.tm.cspirit.client.gui.base.ContainerScreenBase;
import com.tm.cspirit.client.gui.base.TextFieldRect;
import com.tm.cspirit.init.InitSounds;
import com.tm.cspirit.inventory.ContainerPresentUnwrapped;
import com.tm.cspirit.main.ChristmasSpirit;
import com.tm.cspirit.packet.PacketWrapPresent;
import com.tm.cspirit.present.PresentConstructor;
import com.tm.cspirit.present.PresentStyle;
import com.tm.cspirit.tileentity.TileEntityPresentUnwrapped;
import net.minecraft.client.util.InputMappings;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.inventory.container.Container;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.StringTextComponent;
import net.minecraft.util.text.TextFormatting;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ScreenPresentUnwrapped extends ContainerScreenBase<ContainerPresentUnwrapped> {

    private final TileEntityPresentUnwrapped present;
    private final PresentConstructor constructor;

    private TextFieldRect toPlayerNameField;
    private ButtonRect dayBtn, styleBtn;

    public ScreenPresentUnwrapped(Container container, PlayerInventory playerInventory, ITextComponent title) {
        super(container, playerInventory, title);
        playerInventoryTitleY = 102;
        constructor = new PresentConstructor();
        this.present = (TileEntityPresentUnwrapped) getTileEntity();
    }

    @Override
    protected void init() {
        super.init();

        if (minecraft != null) {

            //To Player Name Field
            toPlayerNameField = new TextFieldRect(minecraft.fontRenderer, getScreenX() + 41, getScreenY() + 45, 94, 20, constructor.getFromPlayerName());
            children.add(toPlayerNameField);

            //Day Button
            dayBtn = addButton(new ButtonRect(getScreenX() + 26, getScreenY() + 63, 50, "Day " + constructor.getActualDay(), (btn) -> cycleDays()));

            //Style Button
            styleBtn = addButton(new ButtonRect(getScreenX() + 100, getScreenY() + 63, 50, constructor.getStyle().getName(), (btn) -> cycleStyles()));

            //Wrap Button
            addButton(new ButtonRect(getScreenX() + 51, getScreenY() + 84, 74, "Wrap Present", (btn) -> wrapPresent()));
        }
    }

    private void cycleDays() {
        constructor.setDay(constructor.getDay() + 1);

        if (constructor.getDay() > 32) {
            constructor.setDay(0);
        }

        dayBtn.setMessage(new StringTextComponent("Day " + constructor.getActualDay()));
    }

    private void cycleStyles() {
        constructor.setStyleIndex(constructor.getStyleIndex() + 1);

        if (constructor.getStyleIndex() > PresentStyle.values().length - 1) {
            constructor.setStyleIndex(0);
        }

        styleBtn.setMessage(new StringTextComponent(constructor.getStyle().getName()));
    }

    private boolean isPresentReady() {

        boolean notEmpty = !present.getInventory().getStackInSlot(0).isEmpty();
        boolean hasToPlayerName = !toPlayerNameField.getText().isEmpty();

        if (!notEmpty) present.getUnitName(player).printMessage(TextFormatting.RED, "The present is empty!");
        if (!hasToPlayerName) present.getUnitName(player).printMessage(TextFormatting.RED, "The present needs a player to go to!");

        return notEmpty && hasToPlayerName;
    }

    private void wrapPresent() {

        if (isPresentReady()) {
            closeScreen();

            Pattern pattern = Pattern.compile("[\u00A7].");
            Matcher matcher = pattern.matcher(player.getDisplayName().getString());
            String cleanedString = matcher.replaceAll("");

            constructor.setFromPlayerName(cleanedString);
            constructor.setToPlayerName(toPlayerNameField.getText());
            ChristmasSpirit.network.sendToServer(new PacketWrapPresent(constructor, present.getPos()));
            player.playSound(InitSounds.PRESENT_WRAP.get(), 1, 1);
        }
    }

    @Override
    protected void drawGuiBackground(MatrixStack matrixStack, int mouseX, int mouseY) {
        toPlayerNameField.render(matrixStack, mouseX, mouseY, 0);
    }

    @Override
    protected void drawGuiForeground(MatrixStack matrixStack, int mouseX, int mouseY) {
        if (minecraft != null) minecraft.fontRenderer.drawString(matrixStack, "To:", getScreenX() + 24, getScreenY() + 49, TEXT_COLOR_GRAY);
    }

    @Override
    protected String getGuiTextureName() {
        return "present_unwrapped";
    }

    @Override
    public boolean keyPressed(int keyCode, int scanCode, int modifiers) {

        InputMappings.Input key = InputMappings.getInputByCode(keyCode, scanCode);

        if (minecraft != null && minecraft.gameSettings.keyBindInventory.isActiveAndMatches(key)) {

            if (toPlayerNameField.isFocused()) {
                return true;
            }
        }

        return super.keyPressed(keyCode, scanCode, modifiers);
    }

    @Override
    public int getGuiSizeY () {
        return 194;
    }
}
